### ISSUE TITLE
Add startup / shutdown methods to Agents in Core and AgentChat

### ISSUE THREAD

**Issue by victordibia** [microsoft/autogen#4705]
## What
Across many usecases, agents might need to provision some resources that take a while to startup (e.g., a db connection, a webbrowser instance, a code executor endpoint, a file handle etc). In some ways similar to how [FastAPI handles startup and shutdown via lifespan events. ](https://fastapi.tiangolo.com/advanced/events/)
Similarly these resources need to be shut down once the agent is done running e.g., close a db connection, destroy a browser instance, close files etc.
This issue proposes adding some protocol spec to Agents in core and agentchat that requires agents to implement startup and shutdown methods that the runtime can call. 

## How 
Some proposed steps toward accomplishing this 

- [x] Design a protocol for startup/shutdown lifecycle methods
- [x] Update the core runtime to honor these methods
- [x] Update core and agentchat agent base agent classes to follow protocol
- [ ] Update example agents like websurfer, assistant agent in agentchat to use this design where appropriate
- [ ] Add documentation and examples for implementing startup/shutdown in custom agents 
- [ ] Consider cases for resource pooling and if the framework should enable this e.g., a browser instance 

This should not be a breaking change for any users of existing core or agentchat agents. However users of custom agents might need to implement these methods?

Related to #4699

**Comment by ekzhu** [comment]
## Proposal:

### Added methods

Add the followings to both AgentChat and Core agent protocols:

```python
async def activate(self) -> None:
  ...

async def deactivate(self) -> None:
  ...

### How are they used

How these two methods are used depends on the agent runtime implementation. For `SingleThreadedAgentRuntime`, each agent is recorded with a flag indicating whether has been activated or not.
 
For simplicity, the initial implementation:

`activate` is called when:

- A message is delivered to the agent instance, and the agent is not activated -- lazy activation. 

`deactivate` is called when:

- The runtime is stopped when any of the stop* methods are called.

**Comment by peterychang** [comment]
A few updates:

We're now using RAII, so any initialization an agent needs should be done in the factory method, not in an activation function. `deactivate` has also been renamed to `close()`, which will get run on every instantiated agent when the conditions described by Eric are met. The default implementation is a no-op.

For now, its only implemented in the `SingleThreadedAgentRuntime` but will need to be extended to the other runtimes at some point in the future. Probably a low priority bit for the time being though.
